File: backend/agents/gap_analyzer.py
import math
from nim_client import embed_texts
from agents.gap_models import GapItem, GapAnalysis, MatchStatus
from agents.jd_models import JDStructured
from agents.resume_models import ResumeStructured


EMBEDDING_MODEL = "nvidia/nv-embedqa-e5-v5"

MATCH_THRESHOLD = 0.70
PARTIAL_THRESHOLD = 0.45


def cosine_similarity(a: list[float], b: list[float]) -> float:
    dot = sum(x * y for x, y in zip(a, b))
    norm_a = math.sqrt(sum(x * x for x in a))
    norm_b = math.sqrt(sum(x * x for x in b))
    if norm_a == 0 or norm_b == 0:
        return 0.0
    return dot / (norm_a * norm_b)


def _build_resume_evidence(resume: ResumeStructured) -> list[str]:
    evidence = []
    evidence.extend(resume.skills)
    for exp in resume.experience:
        for bullet in exp.bullets:
            evidence.append(bullet)
    for proj in resume.projects:
        evidence.append(f"{proj.name}: {proj.description} ({', '.join(proj.technologies)})")
    return evidence


def analyze_gaps(jd: JDStructured, resume: ResumeStructured) -> GapAnalysis:
    requirements: list[tuple[str, str]] = []
    for skill in jd.hard_skills:
        requirements.append((skill, "hard_skill"))
    for skill in jd.soft_skills:
        requirements.append((skill, "soft_skill"))
    for kw in jd.domain_keywords:
        requirements.append((kw, "domain_keyword"))
    for bait in jd.ats_bait:
        requirements.append((bait, "ats_bait"))

    if not requirements:
        return GapAnalysis(
            overall_match=1.0,
            matched_count=0,
            partial_count=0,
            missing_count=0,
            total_requirements=0,
            gaps=[],
        )

    resume_evidence = _build_resume_evidence(resume)
    if not resume_evidence:
        return GapAnalysis(
            overall_match=0.0,
            matched_count=0,
            partial_count=0,
            missing_count=len(requirements),
            total_requirements=len(requirements),
            gaps=[
                GapItem(
                    requirement=req,
                    status=MatchStatus.MISSING,
                    similarity_score=0.0,
                    matched_evidence=None,
                    category=cat,
                )
                for req, cat in requirements
            ],
        )

    all_texts = [r[0] for r in requirements] + resume_evidence
    embeddings = embed_texts(all_texts, model=EMBEDDING_MODEL)

    req_embeddings = embeddings[: len(requirements)]
    res_embeddings = embeddings[len(requirements) :]

    gaps: list[GapItem] = []
    for i, (req_text, category) in enumerate(reqlications):
        best_score = 0.0
        best_evidence = None
        for j, res_text in enumerate(resume_evidence):
            score = cosine_similarity(req_embeddings[i], res_embeddings[j])
            if score > best_score:
                best_score = score
                best_evidence = res_text

        if best_score >= MATCH_THRESHOLD:
            status = MatchStatus.MATCHED
        elif best_score >= PARTIAL_THRESHOLD:
            status = MatchStatus.PARTIAL
        else:
            status = MatchStatus.MISSING

        gaps.append(
            GapItem(
                requirement=req_text,
                status=status,
                similarity_score=round(best_score, 4),
                matched_evidence=best_evidence,
                category=category,
            )
        )

    gaps.sort(key=lambda g: g.similarity_score)

    matched = sum(1 for g in gaps if g.status == MatchStatus.MATCHED)
    partial = sum(1 for g in gaps if g.status == MatchStatus.PARTIAL)
    missing = sum(1 for g in gaps if g.status == MatchStatus.MISSING)
    total = len(requirements)
    overall = (matched + partial * 0.5) / total if total > 0 else 0.0

    return GapAnalysis(
        overall_match=round(overall, 4),
        matched_count=matched,
        partial_count=partial,
        missing_count=missing,
        total_requirements=total,
        gaps=gaps,
    )

File: proteus-next/src/lib/agents/gap-analyzer.ts
import type { JDStructured, ResumeStructured, GapAnalysis, GapItem } from "../../types";

function buildResumeEvidence(resume: ResumeStructured): string[] {
  const evidence: string[] = [...resume.skills];
  for (const exp of resume.experience) {
    for (const bullet of exp.bullets) {
      evidence.push(bullet);
    }
  }
  for (const proj of resume.projects) {
    evidence.push(`${proj.name}: ${proj.description} (${proj.technologies.join(", ")})`);
  }
  return evidence;
}

function buildEvidenceLookup(evidence: string[]): { texts: string[]; lower: string[]; words: Set<string> } {
  const lower = evidence.map((e) => e.toLowerCase());
  const words = new Set<string>();
  for (const l of lower) {
    for (const w of l.split(/[^a-z0-9]+/)) {
      if (w.length >= 2) words.add(w);
    }
  }
  return { texts: evidence, lower, words };
}

function matchScore(req: string, lookup: { texts: string[]; lower: string[]; words: Set<string> }): { score: number; evidence: string | null; status: "matched" | "partial" | "missing" } {
  const reqLower = req.toLowerCase().trim();

  // 1. Exact substring match
  for (let i = 0; i < lookup.lower.length; i++) {
    if (lookup.lower[i].includes(reqLower) || reqLower.includes(lookup.lower[i])) {
      return { score: 1.0, evidence: lookup.texts[i], status: "matched" };
    }
  }

  // 2. Single-word exact match
  const reqWords = reqLower.split(/[^a-z0-9]+/).filter((w) => w.length >= 2);
  for (const rw of reqWords) {
    if (lookup.words.has(rw)) {
      for (let i = 0; i < lookup.lower.length; i++) {
        if (lookup.lower[i].split(/[^a-z0-9]+/).includes(rw)) {
          return { score: 0.95, evidence: lookup.texts[i], status: "matched" };
        }
      }
    }
  }

  // 3. Multi-word: all words found
  if (reqWords.length > 1) {
    const allFound = reqWords.every((w) => lookup.words.has(w));
    if (allFound) {
      const evidenceLine = lookup.texts.find((t) => reqWords.every((w) => t.toLowerCase().includes(w))) || null;
      return { score: 0.85, evidence: evidenceLine, status: "matched" };
    }
  }

  // 4. Partial: at least one word matches
  const matchedWords = reqWords.filter((w) => lookup.words.has(w));
  if (matchedWords.length > 0 && matchedWords.length < reqWords.length) {
    return { score: matchedWords.length / reqWords.length, evidence: null, status: "partial" };
  }

  return { score: 0, evidence: null, status: "missing" };
}

export async function analyzeGaps(
  jd: JDStructured,
  resume: ResumeStructured
): Promise<GapAnalysis> {
  const requirements: Array<[string, string]> = [];
  const seen = new Set<string>();
  for (const skill of jd.hard_skills) {
    const key = skill.toLowerCase().trim();
    if (!seen.has(key)) { seen.add(key); requirements.push([skill, "hard_skill"]); }
  }
  for (const skill of jd.soft_skills) {
    const key = skill.toLowerCase().trim();
    if (!seen.has(key)) { seen.add(key); requirements.push([skill, "soft_skill"]); }
  }
  for (const kw of jd.domain_keywords) {
    const key = kw.toLowerCase().trim();
    if (!seen.has(key)) { seen.add(key); requirements.push([kw, "domain_keyword"]); }
  }
  for (const bait of jd.ats_bait) {
    const key = bait.toLowerCase().trim();
    if (!seen.has(key)) { seen.add(key); requirements.push([bait, "ats_bait"]); }
  }

  if (requirements.length === 0) {
    return {
      overall_match: 1.0,
      matched_count: 0,
      partial_count: 0,
      missing_count: 0,
      total_requirements: 0,
      gaps: [],
    };
  }

  const resumeEvidence = buildResumeEvidence(resume);
  if (resumeEvidence.length === 0) {
    return {
      overall_match: 0.0,
      matched_count: 0,
      partial_count: 0,
      missing_count: requirements.length,
      total_requirements: requirements.length,
      gaps: requirements.map(([req, cat]) => ({
        requirement: req,
        status: "missing" as const,
        similarity_score: 0.0,
        matched_evidence: null,
        category: cat as "hard_skill" | "soft_skill" | "domain_keyword" | "ats_bait",
      })),
    };
  }

  const lookup = buildEvidenceLookup(resumeEvidence);

  const gaps: GapItem[] = [];
  for (const [reqText, category] of requirements) {
    const result = matchScore(reqText, lookup);
    gaps.push({
      requirement: reqText,
      status: result.status,
      similarity_score: Math.round(result.score * 10000) / 10000,
      matched_evidence: result.evidence,
      category: category as "hard_skill" | "soft_skill" | "domain_keyword" | "ats_bait",
    });
  }

  gaps.sort((a, b) => a.similarity_score - b.similarity_score);

  const matched = gaps.filter((g) => g.status === "matched").length;
  const partial = gaps.filter((g) => g.status === "partial").length;
  const missing = gaps.filter((g) => g.status === "missing").length;
  const total = requirements.length;
  const overall = total > 0 ? (matched * 1.0 + partial * 0.6) / total : 0;

  return {
    overall_match: Math.round(overall * 10000) / 10000,
    matched_count: matched,
    partial_count: partial,
    missing_count: missing,
    total_requirements: total,
    gaps,
  };
}
